### ISSUE TITLE
[Bug]: Fatal Crash when unpacking Unity Crunch

### ISSUE THREAD

**Issue by ds5678** [AssetRipper/AssetRipper#81]
### Are you on the latest version of AssetRipper?

Yes, I'm on the latest alpha-build of AssetRipper.

### Which release are you using?

Console Version Windows x64

### Which Unity version did this occur on?

Not applicable

### Is the game Mono or IL2Cpp?

Not Applicable

### Describe the issue.

This bug occurs when the ripper tries to unpack a texture compressed with Unity Crunch. It is a fatal exception which bypasses the try catch statement in the console version.

This is directly caused by the transition to Texture2DDecoderNative. There must be some flaw there because this bug doesn't occur when using the old unity crunch binary. 

In order to find the issue, the [code](https://github.com/ds5678/AssetRipper/tree/master/Texture2DDecoderNative/unitycrunch) should be compared with the [original source](https://github.com/Unity-Technologies/crunch). It is also important to verify that d1efa2dfecd16d1a58cbfed75f13de6479cbfd26 and 10e1c5bc851d5ab38345d3702eefd19758501f19 are not the cause of this issue.

### Relevant log output

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at Texture2DDecoder.TextureDecoder.UnpackUnityCrunch(Void*, UInt32, Void* ByRef, UInt32 ByRef)
--------------------------------
   at Texture2DDecoder.TextureDecoder.UnpackUnityCrunch(Byte[])
   at AssetRipper.Library.Exporters.Textures.TextureConverter.DecompressUnityCRN(Byte[], Byte[] ByRef)
   at AssetRipper.Library.Exporters.Textures.TextureConverter.DecompressCrunch(AssetRipper.Core.Classes.Texture2D.TextureFormat, Int32, Int32, AssetRipper.Core.Parser.Files.UnityVersion, Byte[])
   at AssetRipper.Library.Exporters.Textures.TextureAssetExporter.ConvertToBitmap(AssetRipper.Core.Classes.Texture2D.TextureFormat, Int32, Int32, AssetRipper.Core.Parser.Files.UnityVersion, Byte[], Int32, Int32, AssetRipper.Library.TextureContainers.KTX.KTXBaseInternalFormat)
   at AssetRipper.Library.Exporters.Textures.TextureAssetExporter.ExportTexture(AssetRipper.Core.Classes.Texture2D.Texture2D, System.IO.Stream, AssetRipper.Library.Configuration.ImageExportFormat)
   at AssetRipper.Library.Exporters.Textures.TextureAssetExporter.Export(AssetRipper.Core.Project.IExportContainer, AssetRipper.Core.Classes.Object.Object, System.String)
   at AssetRipper.Core.Structure.Collections.AssetExportCollection.Export(AssetRipper.Core.Project.ProjectAssetContainer, System.String)
   at AssetRipper.Core.Project.ProjectExporter.Export(AssetRipper.Core.Structure.GameCollection, System.Collections.Generic.IEnumerable`1<AssetRipper.Core.Parser.Files.SerializedFiles.SerializedFile>, AssetRipper.Core.Configuration.CoreConfiguration)
   at AssetRipper.Core.Project.ProjectExporter.Export(AssetRipper.Core.Structure.GameCollection, AssetRipper.Core.Configuration.CoreConfiguration)
   at AssetRipper.Core.Structure.GameCollection.Export(AssetRipper.Core.Configuration.CoreConfiguration)
   at AssetRipper.Core.Structure.GameStructure.GameStructure.Export(AssetRipper.Core.Configuration.CoreConfiguration)
   at AssetRipper.Library.Ripper.ExportProject(System.String)
   at AssetRipper.Console.Program.Run(Options)
   at AssetRipper.Console.Program+<>c.<Main>b__2_0(Options)
   at CommandLine.ParserResultExtensions.WithParsed[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](CommandLine.ParserResult`1<System.__Canon>, System.Action`1<System.__Canon>)
   at AssetRipper.Console.Program.Main(System.String[])
```

**Comment by ds5678** [comment]
Note: it is likely that decompressing normal crunched textures will have the same effect. The [crunch code](https://github.com/ds5678/AssetRipper/tree/master/Texture2DDecoderNative/crunch) should be compared with its [original source](https://github.com/BinomialLLC/crunch) too.

**Comment by DatTaiwanGuy** [comment]
Utinyripper seems to be okay with unpacking and decompressing Unity Crunched textures than the updated AssetRipper.

**Comment by ds5678** [comment]
@DatTaiwanGuy That is because AssetRipper has to support Mac and Linux in addition to Windows, so we can't use the native binaries that uTinyRipper does.

**Comment by JavidPack** [comment]
I also experienced this issue, and thought a comment might show support for the bug. I don't really have any other relevant info, but it is strange that the whole program will crash for an exception like this. I wonder, is there a way to skip the specific file in a subsequent attempt? It seems the console log prior to the exception shows the name of the last successful exported asset, not the one that actually errored, so I'm not even sure which file caused the error.

I am on the latest CI build from cda185d

**Comment by ds5678** [comment]
A temporary fix for Windows was implemented in 21fd719db34f54beea0c54a99c673aaf5547cafc. Mac and Linux users will likely still be able to experience this issue. Future plans include porting these native dependencies to C# for platform independence.

**Comment by ds5678** [comment]
Superceded by #448
